### PR TITLE
Cl/handle search infrastructure exceptions

### PR DESCRIPTION
**Pull request by CathLass**
Added and modified tests to better define and test behaviour
Added a response status to the application layer
Some small tidy-ups like adding a models folder which now makes the PR look massive - sorry about that :(

### CODE REVIEW

File: Dfe.Data.SearchPrototype/SearchForEstablishments/Models/EstablishmentStatusCode.cs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Dfe.Data.SearchPrototype.SearchForEstablishments.Models;
+
+public enum EstablishmentStatusCode
+{
+    Closed = 0,
+    Open = 1,
+    Unknown
+}
```
Comment (CathLass) [inline Dfe.Data.SearchPrototype/SearchForEstablishments/Models/EstablishmentStatusCode.cs:8]: This is just a rename from StatusCode - not new code

File: Dfe.Data.SearchPrototype/SearchForEstablishments/SearchResponseStatus.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Dfe.Data.SearchPrototype.SearchForEstablishments;
+
+public enum SearchResponseStatus
+{
+    Success,
+    NoResults,
+    InvalidRequest,
+    SearchServiceError,
+    DataFormatError
+}
```
Comment (CathLass) [inline Dfe.Data.SearchPrototype/SearchForEstablishments/SearchResponseStatus.cs:8]: This is the actual main thing I've added - this response status and the setting of it by the use-case and the mapper

File: Dfe.Data.SearchPrototype/SearchForEstablishments/SearchResponseStatus.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Dfe.Data.SearchPrototype.SearchForEstablishments;
+
+public enum SearchResponseStatus
+{
+    Success,
+    InvalidRequest,
+    SearchServiceError,
+    DataFormatError
```
Comment (RogerHowellDfE) [inline Dfe.Data.SearchPrototype/SearchForEstablishments/SearchResponseStatus.cs]: `DataFormatError` doesn't currently appear to be used.

Out of curiosity, what is the intention for this case? Data is returned from the search API, but the mapper fails to transform/map it out of the Adapter, perhaps? (if so, maybe we need to add try/catch or some other logic to return this?)

Presumably this repo/context doesn't care about the transport encoding (JSON/XML/binary/something else) and everything this side of the adapter is dealing with de-serialised C# objects -- assuming this is correct, I can't quite get my head around what `Format` is considered erroneous at this point.

File: Dfe.Data.SearchPrototype/Infrastructure/Tests/CognitiveSearchServiceAdapterTests.cs [hunk 4]
```diff
@@ -49,14 +50,18 @@ public async Task Search_With_Valid_SearchContext_Returns_Configured_Results()
...
+
         // arrange
         ISearchServiceAdapter cognitiveSearchServiceAdapter =
             CreateServiceAdapterWith(
-                SearchServiceTestDouble.MockSearchService("SearchKeyword", "TargetCollection"),
-                SearchOptionsFactoryTestDouble.MockForDefaultResult(),
-                AzureSearchResponseToSearchResultsMapperTestDouble.MockDefaultMapper());
+                mockService,
+                mockSearchOptionsFactory,
+                mockMapper);
```
Comment (RogerHowellDfE) [inline Dfe.Data.SearchPrototype/Infrastructure/Tests/CognitiveSearchServiceAdapterTests.cs:64]: Possibly a comment for a different PR (so feel free to disregard / resolve this comment without changes), but I would be interested to understand the rationale for having an additional `CreateServiceAdapterWith` static method, as opposed to in-lining it (and removing a level of indirection) like this:

```csharp
        // arrange
        ISearchServiceAdapter cognitiveSearchServiceAdapter =
            new CognitiveSearchServiceAdapter<Establishment>(mockService, mockSearchOptionsFactory, mockMapper);
```

Could just be a historical thing where an older version was more complex, perhaps? Or just a style thing? Or to help with the generics side of things maybe?

_(regardless, for me this is just a curiosity and possibly style observation -- certainly nothing I want to get hung up on or block the PR for! 👍)_

**Review by RogerHowellDfE — APPROVED** [review]
With the caveat of my comment about the currently unused `SearchResponseStatus.DataFormatError` (personally I'd not want to block the PR based just on this unused value) and caution re: I don't _quite_ have a full mental model yet of how everything fits together, it's a ✅ from me 👍 

- the commits seem to follow a logical progression
- the tests _seem_ to be testing what they say to be testing
- overall this seems a good/reasonable improvement

A few nitpicky comments _if_ we were to refine the PR (i.e., not anything I'd want to block the PR on):

- Seems to be a few nullability mismatches floating around (parameters marked as non-nullable then a robust additional runtime check for null causing warnings, and a few places in the tests use `!` to override the nullability declarations)
- Possibly split the PR into refactoring changes (renaming/moving) vs functionality (behaviour) changes?

File: Dfe.Data.SearchPrototype/SearchForEstablishments/SearchResponseStatus.cs (hunk 3 above)
Comment (CathLass) [reply Dfe.Data.SearchPrototype/SearchForEstablishments/SearchResponseStatus.cs]: Yes, I added it thinking I was going to use it if the mapping went wrong and then didn't

File: Dfe.Data.SearchPrototype/Infrastructure/Tests/CognitiveSearchServiceAdapterTests.cs (hunk 4 above)
Comment (CathLass) [reply Dfe.Data.SearchPrototype/Infrastructure/Tests/CognitiveSearchServiceAdapterTests.cs:64]: Don't know about why it's there, so I kept as is

**Pull request merged**